Bug in parsing timestamptz

I believe the following error is encountered when parsing the timestamptz (timestamp with timezone) from Redshift.

```Py4JJavaError: An error occurred while calling o227.collectToPython.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 1.0 failed 4 times, most recent failure: Lost task 0.3 in stage 1.0 (TID 323, 10-81-19-120-uswest2adevc.dev.yelpcorp.com, executor 1): java.lang.NumberFormatException: For input string: "29+00"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Integer.parseInt(Integer.java:580)
	at java.lang.Integer.parseInt(Integer.java:615)
	at java.sql.Timestamp.valueOf(Timestamp.java:259)
	at io.github.spark_redshift_community.spark.redshift.Conversions$$anonfun$1$$anonfun$apply$11.apply(Conversions.scala:107)
	at io.github.spark_redshift_community.spark.redshift.Conversions$$anonfun$1$$anonfun$apply$11.apply(Conversions.scala:107)
	at io.github.spark_redshift_community.spark.redshift.Conversions$$anonfun$createRowConverter$1.apply(Conversions.scala:127)
	at io.github.spark_redshift_community.spark.redshift.Conversions$$anonfun$createRowConverter$1.apply(Conversions.scala:115)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.next(FileScanRDD.scala:104)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$13$$anon$1.hasNext(WholeStageCodegenExec.scala:636)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$13$$anon$1.hasNext(WholeStageCodegenExec.scala:636)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:255)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:247)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:121)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1889)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1877)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1876)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1876)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)
	at scala.Option.foreach(Option.scala:257)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:926)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2110)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2059)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2048)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:737)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2061)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2082)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2101)
	at org.apache.spark.sql.execution.SparkPlan.executeTake(SparkPlan.scala:365)
	at org.apache.spark.sql.execution.CollectLimitExec.executeCollect(limit.scala:38)
	at org.apache.spark.sql.Dataset$$anonfun$collectToPython$1.apply(Dataset.scala:3257)
	at org.apache.spark.sql.Dataset$$anonfun$collectToPython$1.apply(Dataset.scala:3254)
	at org.apache.spark.sql.Dataset$$anonfun$53.apply(Dataset.scala:3364)
	at org.apache.spark.sql.execution.SQLExecution$$anonfun$withNewExecutionId$1.apply(SQLExecution.scala:78)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:125)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:73)
	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3363)
	at org.apache.spark.sql.Dataset.collectToPython(Dataset.scala:3254)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:238)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NumberFormatException: For input string: "29+00"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Integer.parseInt(Integer.java:580)
	at java.lang.Integer.parseInt(Integer.java:615)
	at java.sql.Timestamp.valueOf(Timestamp.java:259)
	at io.github.spark_redshift_community.spark.redshift.Conversions$$anonfun$1$$anonfun$apply$11.apply(Conversions.scala:107)
	at io.github.spark_redshift_community.spark.redshift.Conversions$$anonfun$1$$anonfun$apply$11.apply(Conversions.scala:107)
	at io.github.spark_redshift_community.spark.redshift.Conversions$$anonfun$createRowConverter$1.apply(Conversions.scala:127)
	at io.github.spark_redshift_community.spark.redshift.Conversions$$anonfun$createRowConverter$1.apply(Conversions.scala:115)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.next(FileScanRDD.scala:104)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$13$$anon$1.hasNext(WholeStageCodegenExec.scala:636)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$13$$anon$1.hasNext(WholeStageCodegenExec.scala:636)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:255)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:247)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:121)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	... 1 more
```

--- a/src/it/scala/io/github/spark_redshift_community/spark/redshift/RedshiftReadSuite.scala
+++ b/src/it/scala/io/github/spark_redshift_community/spark/redshift/RedshiftReadSuite.scala
@@ -215,6 +215,21 @@ class RedshiftReadSuite extends IntegrationSuiteBase {
     }
   }
 
+  test("test timestamptz parsing") {
+    withTempRedshiftTable("luca_test_timestamptz_spark_redshift") { tableName =>
+      conn.createStatement().executeUpdate(
+        s"CREATE TABLE $tableName (x timestamptz)"
+      )
+      conn.createStatement().executeUpdate(
+        s"INSERT INTO $tableName VALUES ('2015-07-03 00:00:00.000 -0300')"
+      )
+
+      checkAnswer(
+        read.option("dbtable", tableName).load(),
+        Seq(Row.apply("2015-07-03 03:00:00.0"))
+      )
+    }
+  }
 
   test("read special double values (regression test for #261)") {
     val tableName = s"roundtrip_special_double_values_$randomSuffix"

--- a/src/main/scala/io/github/spark_redshift_community/spark/redshift/Conversions.scala
+++ b/src/main/scala/io/github/spark_redshift_community/spark/redshift/Conversions.scala
@@ -18,6 +18,8 @@ package io.github.spark_redshift_community.spark.redshift
 
 import java.sql.Timestamp
 import java.text.{DecimalFormat, DecimalFormatSymbols, SimpleDateFormat}
+import java.time.{DateTimeException, LocalDateTime, ZonedDateTime}
+import java.time.format.DateTimeFormatter
 import java.util.Locale
 
 import org.apache.spark.sql.catalyst.InternalRow
@@ -74,6 +76,25 @@ private[redshift] object Conversions {
   }
 
   /**
+    * From the DateTimeFormatter docs (Java 8):
+    * "A formatter created from a pattern can be used as many times as necessary, it is immutable and is thread-safe."
+    */
+  private val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss[.SSS][.SS][.S][ X]")
+
+  def parseRedshiftTimestamp(s: String): Timestamp = {
+    val temporalAccessor = formatter.parse(s)
+
+    try {
+      Timestamp.valueOf(ZonedDateTime.from(temporalAccessor).toLocalDateTime)
+    }
+    catch {
+      // Case timestamp without timezone
+      case e: DateTimeException =>
+        Timestamp.valueOf(LocalDateTime.from(temporalAccessor))
+    }
+  }
+
+  /**
    * Return a function that will convert arrays of strings conforming to the given schema to Rows.
    *
    * Note that instances of this function are NOT thread-safe.
@@ -104,7 +125,7 @@ private[redshift] object Conversions {
         case LongType => (data: String) => java.lang.Long.parseLong(data)
         case ShortType => (data: String) => java.lang.Short.parseShort(data)
         case StringType => (data: String) => data
-        case TimestampType => (data: String) => Timestamp.valueOf(data)
+        case TimestampType => (data: String) => parseRedshiftTimestamp(data)
         case _ => (data: String) => data
       }
     }

--- a/src/main/scala/io/github/spark_redshift_community/spark/redshift/Conversions.scala
+++ b/src/main/scala/io/github/spark_redshift_community/spark/redshift/Conversions.scala
@@ -77,7 +77,8 @@ private[redshift] object Conversions {
 
   /**
     * From the DateTimeFormatter docs (Java 8):
-    * "A formatter created from a pattern can be used as many times as necessary, it is immutable and is thread-safe."
+    * "A formatter created from a pattern can be used as many times as necessary,
+    * it is immutable and is thread-safe."
     */
   private val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss[.SSS][.SS][.S][ X]")
 

--- a/src/main/scala/io/github/spark_redshift_community/spark/redshift/Conversions.scala
+++ b/src/main/scala/io/github/spark_redshift_community/spark/redshift/Conversions.scala
@@ -33,6 +33,13 @@ import org.apache.spark.sql.types._
 private[redshift] object Conversions {
 
   /**
+    * From the DateTimeFormatter docs (Java 8):
+    * "A formatter created from a pattern can be used as many times as necessary,
+    * it is immutable and is thread-safe."
+    */
+  private val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss[.SSS][.SS][.S][X]")
+
+  /**
    * Parse a boolean using Redshift's UNLOAD bool syntax
    */
   private def parseBoolean(s: String): Boolean = {
@@ -75,18 +82,12 @@ private[redshift] object Conversions {
     new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS")
   }
 
-  /**
-    * From the DateTimeFormatter docs (Java 8):
-    * "A formatter created from a pattern can be used as many times as necessary,
-    * it is immutable and is thread-safe."
-    */
-  private val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss[.SSS][.SS][.S][X]")
-
   def parseRedshiftTimestamp(s: String): Timestamp = {
     val temporalAccessor = formatter.parse(s)
 
     try {
-      Timestamp.valueOf(ZonedDateTime.from(temporalAccessor).toLocalDateTime)
+      // timestamptz
+      Timestamp.from(ZonedDateTime.from(temporalAccessor).toInstant)
     }
     catch {
       // Case timestamp without timezone

--- a/src/main/scala/io/github/spark_redshift_community/spark/redshift/Conversions.scala
+++ b/src/main/scala/io/github/spark_redshift_community/spark/redshift/Conversions.scala
@@ -80,7 +80,7 @@ private[redshift] object Conversions {
     * "A formatter created from a pattern can be used as many times as necessary,
     * it is immutable and is thread-safe."
     */
-  private val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss[.SSS][.SS][.S][ X]")
+  private val formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss[.SSS][.SS][.S][X]")
 
   def parseRedshiftTimestamp(s: String): Timestamp = {
     val temporalAccessor = formatter.parse(s)

--- a/src/test/scala/io/github/spark_redshift_community/spark/redshift/ConversionsSuite.scala
+++ b/src/test/scala/io/github/spark_redshift_community/spark/redshift/ConversionsSuite.scala
@@ -62,9 +62,10 @@ class ConversionsSuite extends FunSuite {
     val rowConverter = createRowConverter(
       StructType(Seq(StructField("timestampWithTimezone", TimestampType))))
 
-    val timestampWithTimezone = "2014-03-01 00:00:01.123+00"
+    // when converting to timestamp, we discard the TZ info.
+    val timestampWithTimezone = "2014-03-01 00:00:01.123-03"
     val expectedTimestampWithTimezoneMillis = TestUtils.toMillis(
-      2014, 2, 1, 0, 0, 1, 123)
+      2014, 1, 28, 19, 0, 1, 123)
 
     val convertedRow = rowConverter(Array(timestampWithTimezone))
     val expectedRow = Row(new Timestamp(expectedTimestampWithTimezoneMillis))

--- a/src/test/scala/io/github/spark_redshift_community/spark/redshift/ConversionsSuite.scala
+++ b/src/test/scala/io/github/spark_redshift_community/spark/redshift/ConversionsSuite.scala
@@ -17,6 +17,7 @@
 package io.github.spark_redshift_community.spark.redshift
 
 import java.sql.Timestamp
+import java.time.{LocalDateTime, ZoneId, ZoneOffset, ZonedDateTime}
 import java.util.Locale
 
 import org.apache.spark.sql.Row
@@ -43,9 +44,9 @@ class ConversionsSuite extends FunSuite {
     // scalastyle:on
 
     val timestampWithMillis = "2014-03-01 00:00:01.123"
+    val expectedTimestampMillis = TestUtils.toMillis(2014, 2, 1, 0, 0, 1, 123)
 
     val expectedDateMillis = TestUtils.toMillis(2015, 6, 1, 0, 0, 0)
-    val expectedTimestampMillis = TestUtils.toMillis(2014, 2, 1, 0, 0, 1, 123)
 
     val convertedRow = convertRow(
       Array("1", "t", "2015-07-01", doubleMin, "1.0", "42",
@@ -64,8 +65,9 @@ class ConversionsSuite extends FunSuite {
 
     // when converting to timestamp, we discard the TZ info.
     val timestampWithTimezone = "2014-03-01 00:00:01.123-03"
+
     val expectedTimestampWithTimezoneMillis = TestUtils.toMillis(
-      2014, 1, 28, 19, 0, 1, 123)
+      2014, 2, 1, 0, 0, 1, 123, "-03")
 
     val convertedRow = rowConverter(Array(timestampWithTimezone))
     val expectedRow = Row(new Timestamp(expectedTimestampWithTimezoneMillis))

--- a/src/test/scala/io/github/spark_redshift_community/spark/redshift/ConversionsSuite.scala
+++ b/src/test/scala/io/github/spark_redshift_community/spark/redshift/ConversionsSuite.scala
@@ -58,6 +58,20 @@ class ConversionsSuite extends FunSuite {
     assert(convertedRow == expectedRow)
   }
 
+  test("Regression test for parsing timestamptz (bug #25 in spark_redshift_community)") {
+    val rowConverter = createRowConverter(
+      StructType(Seq(StructField("timestampWithTimezone", TimestampType))))
+
+    val timestampWithTimezone = "2014-03-01 00:00:01.123 -0300"
+    val expectedTimestampWithTimezoneMillis = TestUtils.toMillis(
+      2014, 2, 1, 0, 0, 1, 123)
+
+    val convertedRow = rowConverter(Array(timestampWithTimezone))
+    val expectedRow = Row(new Timestamp(expectedTimestampWithTimezoneMillis))
+
+    assert(convertedRow == expectedRow)
+  }
+
   test("Row conversion handles null values") {
     val convertRow = createRowConverter(TestUtils.testSchema)
     val emptyRow = List.fill(TestUtils.testSchema.length)(null).toArray[String]

--- a/src/test/scala/io/github/spark_redshift_community/spark/redshift/ConversionsSuite.scala
+++ b/src/test/scala/io/github/spark_redshift_community/spark/redshift/ConversionsSuite.scala
@@ -62,7 +62,7 @@ class ConversionsSuite extends FunSuite {
     val rowConverter = createRowConverter(
       StructType(Seq(StructField("timestampWithTimezone", TimestampType))))
 
-    val timestampWithTimezone = "2014-03-01 00:00:01.123 -0300"
+    val timestampWithTimezone = "2014-03-01 00:00:01.123+00"
     val expectedTimestampWithTimezoneMillis = TestUtils.toMillis(
       2014, 2, 1, 0, 0, 1, 123)
 

--- a/src/test/scala/io/github/spark_redshift_community/spark/redshift/TestUtils.scala
+++ b/src/test/scala/io/github/spark_redshift_community/spark/redshift/TestUtils.scala
@@ -17,7 +17,8 @@
 package io.github.spark_redshift_community.spark.redshift
 
 import java.sql.{Date, Timestamp}
-import java.util.{Calendar, Locale}
+import java.time.ZoneId
+import java.util.{Calendar, Locale, TimeZone}
 
 import org.apache.spark.sql.Row
 import org.apache.spark.sql.types._
@@ -84,10 +85,12 @@ object TestUtils {
       hour: Int,
       minutes: Int,
       seconds: Int,
-      millis: Int = 0): Long = {
+      millis: Int = 0,
+      timeZone: String = null): Long = {
     val calendar = Calendar.getInstance()
     calendar.set(year, zeroBasedMonth, date, hour, minutes, seconds)
     calendar.set(Calendar.MILLISECOND, millis)
+    if (timeZone != null) calendar.setTimeZone(TimeZone.getTimeZone(ZoneId.of(timeZone)))
     calendar.getTime.getTime
   }
 

--- a/version.sbt
+++ b/version.sbt
@@ -1,1 +1,1 @@
-version in ThisBuild := "4.0.0-preview20190730"
+version in ThisBuild := "4.0.0-preview20190813"

